overlord/snapshotstate/backend/backend.go

```diff
@@ -0,0 +1,220 @@
+package backend
+
+import (
+	"archive/zip"
+	"crypto"
+	"encoding/json"
+	"fmt"
+	"hash"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+)
+
+const (
+	archiveName  = "archive.tgz"
+	metadataName = "meta.json"
+	metaHashName = "meta.sha3_384"
+
+	userArchivePrefix = "user/"
+	userArchiveSuffix = ".tgz"
```

Comment:
this is the name of the thing inside, not the filename (which is something ending in `zip`, as you saw below).

Snapshots are zips of tars + metadata. They're zips because tar needs to know the size of the thing upfront, and we stream the tars in.